### ISSUE TITLE
Best way to deploy to web host running nginx and node

### ISSUE THREAD

**Issue by timothystewart6** [angular-fullstack/generator-angular-fullstack#1641]
Hi.  I am curious if there is any documentation on how to deploy to a web server running nginx and node?   I have it mostly working however it seems like the routes need to be adjusted, fontawesome and parts of bootstrap are not working, etc... then changing those routes would break when serving it local.

So far, I have the backend running with PM2 and the front end being served using nginx however I am not sure if it is working properly since none of the routes seems to work.  Any help is appreciated.

**Comment by timothystewart6** [comment]
I'd be willing to write the documentation too, just looking some some guidance from someone who has experience with this.

**Comment by Koslun** [comment]
Need more information on what steps you've taken and maybe a print of the actual errors to really say what exactly could fix whatever problem you currently have.

Have you for instance built the project with "grunt build", "gulp build" or simply running "grunt" or "gulp"? And did you then export the NODE_ENV to production before running it with pm2 (export NODE_ENV=production)? What OS you are deploying it on might also be a factor for any future or current errors.

For tutorials I would however recommend digital ocean, at the very least they do have some tutorials on how to configure and setup nginx with node on ubuntu: https://www.digitalocean.com/community/tutorials/how-to-set-up-a-node-js-application-for-production-on-ubuntu-14-04. They also have some other similar tutorials concerning deployment with nginx and node.

What I've generally seen recommended is to do a reverse-proxy using nginx to serve the backend routes, while many also use and recommend nginx for serving the frontend there is very little performance difference now as far as I understood it. Meaning you could also serve the frontend using the same reverse-proxy configuration. Either way the backend of this generator is configured to actually try to do that, i.e. serve the frontend when queried.

**Comment by timothystewart6** [comment]
Thanks for the response. Sorry I should have been more clear on what is working and what is not.
* reverse proxy is set up per digital ocean
* NODE_ENV is set to production
* web host is linux
* pm2 is working fine, api is up and serving requests
* nginx is up, serving static files
* gulp build to build the server and client
* host files look like /var/www/project/dist/server and /var/www/project/dist/client

Immediately upon browsing to login I get font-awesome and socket.io errors:
```
Failed to load resource: the server responded with a status of 404 (Not Found) http://192.168.0.77/socket.io-client/?EIO=3&transport=polling&t=LBsyg6n 
Failed to load resource: the server responded with a status of 404 (Not Found) http://192.168.0.77/bower_components/bootstrap-sass-official/vendor/assets/fonts/bootstrap/glyphicons-halflings-regular.ttf 
Failed to load resource: the server responded with a status of 404 (Not Found) http://192.168.0.77/socket.io-client/?EIO=3&transport=polling&t=LBsygJC Failed to load resource: the server responded with a status of 404 (Not Found)
GET http://192.168.0.77/socket.io-client/?EIO=3&transport=polling&t=LBsuBs2 404 (Not Found)
GET http://192.168.0.77/bower_components/font-awesome/fonts/fontawesome-webfont.woff2?v=4.4.0
```
I was getting
Since this is minified, it shouldn't be looking at bower_components, should it?

I also get errors when trying to route to the /api path.  I can fix this though with a route although now I need to adjust it in all new routes too.
```
GET http://192.168.0.77/socket.io-client/?EIO=3&transport=polling&t=LBsvj6a 404 (Not Found)
```

I can fix the fontawesome and bootstrap 404s by running bower install on the server however it should be pulling this from the minified vendor scripts.

So to recap, I think I have my environment set up properly however I am not sure of the adjustments I need to make to the template for it to run.  I can probably solve most with modifying the routes but it seems like the GETs for resources from bower_components is a bug - although I could be wrong.  Thanks,

**Comment by Koslun** [comment]
Yeah, you should definitely not `bower install` or `npm install` in the dist directory, everything should already be there after` gulp build`. It would be if you haven't run ahead of `gulp build`.

As you say there might be a bug with the gulp configuration, bower_components should only be loaded if you yourself manually insert the dependency into `index.html`, rather than letting the gulp plugin do it for you. There also seems to be an issue already created for font-awesome not loading using gulp so probably related, see #1602. I don't know if it actually does solve everything but it might lend more insight into how to fix the gulp script manually yourself. Though as you say, you could also possibly do some specific routes in nginx or bower install. I would however rather just include the minified versions manually in the index.html.

Gulp was labeled experimental until very recently so I would consider it generally a bit less stable than the grunt configuration.

Seeing the 502 for `/api paths` I would say that the reverse proxy isn't working yet for the API and you thus need to change the nginx route configuration, assuming that you've verified that the node API is working as expected.

Unsure about the socket errors, have not used sockets much yet myself so don't quite know how they're configured.

**Comment by timothystewart6** [comment]
I am running gulp build.  The wiredep task should be taking care of dependecy wiring up in the index.html, as well as pointing to vendor scripts since a gulp build minifies, mangles, and uglifies.  That's too bad about font-awesome, bootstrap, and gulp.  If I figure it out I can open a PR.  I thought what I am trying to do is pretty typical for most projects that aren't hosting on heroku or openshift and that someone out there had some insight on what needs to change in order to host the api and files on the same server.
